gcp_variant_transforms/beam_io/vcf_parser.py

```diff
@@ -0,0 +1,449 @@
+
+  def __ge__(self, other):
+    return other <= self
+
+  def __ne__(self, other):
+    return not self == other
+
+  def __repr__(self):
+    return ', '.join(
+        [str(s) for s in [self.name, self.genotype, self.phaseset, self.info]])
+
+
+class VcfParser(object):
+  """Base abstract class for defining a vcf file parser.
+
+  Derived classes must implement two methods:
+    _init_with_header: must initial parser with given header lines (List[str]).
+    _get_record: give a line of vcf file, returns a Variant object.
+
+  Objects of the derived classed will be an iterator of records:
+
+    record_iterator = DerivedVcfParser(...)
+    for record in record_iterator:
+      yield record
+  """
+
+  def __init__(self,
+               file_name,  # type: str
+               range_tracker,  # type: range_trackers.OffsetRangeTracker
+               file_pattern,  # type: str
+               compression_type,  # type: str
+               allow_malformed_records,  # type: bool
+               representative_header_lines=None,  # type:  List[str]
+               **kwargs  # type: **str
+              ):
+    # type: (...) -> None
+    # If `representative_header_lines` is given, header lines in `file_name`
+    # are ignored; refer to _process_header_lines() logic.
+    self._header_lines = []
+    self._representative_header_lines = representative_header_lines
+    self._current_line = None
+    self._file_name = file_name
+    self._allow_malformed_records = allow_malformed_records
+
+    self._header_processed = False
+    text_source = textio._TextSource(
+        file_pattern,
+        0,  # min_bundle_size
+        compression_type,
+        True,  # strip_trailing_newlines
+        coders.StrUtf8Coder(),  # coder
+        validate=False,
+        header_processor_fns=(lambda x: x.startswith('#'),
+                              self._process_header_lines),
+        **kwargs)
+
+    self._text_lines = text_source.read_records(self._file_name,
+                                                range_tracker)
+
+  def _process_header_lines(self, header_lines):
+    """ This method will be automatically called by textio._TextSource()."""
+    if self._representative_header_lines:
+      # Replace header lines with given representative header lines.
+      # We need to keep the last line of the header from the file because it
+      # contains the sample IDs, which is unique per file.
+      self._init_with_header(self._representative_header_lines +
+                             header_lines[-1:])
+    else:
+      self._init_with_header(header_lines)
+
+  def next(self):
+    for text_line in self._text_lines:
+      return self._get_record(text_line)
+    raise StopIteration
+
+  def __iter__(self):
+    return self
+
+  def _init_with_header(self, header_lines):
+    # type: (List[str]) -> None
+    """ This method will be called by _process_header_lines()."""
+    raise NotImplementedError
+
+  def _get_record(self, line):
```

Comment:
s/line/data_line may be a better name here (but I dont feel strongly about it).
...it may be obvious, but this the body of this method should only care about parsing data lines and not header lines, so renaming the variable makes it even more clear.
and 'data lines' terminology is from VCF spec: http://samtools.github.io/hts-specs/VCFv4.3.pdf